src/Components/Web.JS/src/Platform/WebAssemblyResourceLoader.ts

```diff
@@ -0,0 +1,195 @@
+      if (!(cachedRequest.url in this.usedCacheKeys)) {
+        await this.cache.delete(cachedRequest);
+      }
+    });
+
+    return Promise.all(deletionPromises);
+  }
+
+  private async tryValidateAndCacheAsync(name: string, cacheKey: string, expectedContentHash: string, networkResponse: Response): Promise<ArrayBuffer> {
+    const responseBuffer = await networkResponse.arrayBuffer();
+
+    // Now is an ideal moment to capture the performance stats for the request, since it
+    // only just completed and is most likely to still be in the buffer. However this is
+    // only done on a 'best effort' basis. Even if we do receive an entry, some of its
+    // properties may be blanked out if it was a CORS request.
+    const performanceEntry = getPerformanceEntry(networkResponse.url);
+    const responseBytes = (performanceEntry && performanceEntry.encodedBodySize) || undefined;
+    this.networkLoads[name] = { responseBytes };
+
+    // crypto.subtle is only enabled on localhost and HTTPS origins
+    // We only write to the cache if we can validate the content hashes
+    if (typeof crypto !== 'undefined' && !!crypto.subtle) {
+      await assertContentHashMatchesAsync(name, responseBuffer, expectedContentHash);
+
+      // Build a custom response object so we can track extra data such as responseBytes
+      // We can't rely on the server sending content-length (ASP.NET Core doesn't by default)
+      await this.cache.put(cacheKey, new Response(responseBuffer, {
+        headers: {
+          'content-type': networkResponse.headers.get('content-type') || '',
+          'content-length': (responseBytes || networkResponse.headers.get('content-length') || '').toString()
+        }
+      }));
+    }
+
+    return responseBuffer;
+  }
+}
+
+function countTotalBytes(loads: LoadLogEntry[]) {
+  return loads.reduce((prev, item) => prev + (item.responseBytes || 0), 0);
+}
+
+function toDataSizeString(byteCount: number) {
+  return `${(byteCount / (1024*1024)).toFixed(2)} MB`;
+}
+
+function getPerformanceEntry(url: string): PerformanceResourceTiming | undefined {
+  if (typeof performance !== 'undefined') {
+    return performance.getEntriesByName(url)[0] as PerformanceResourceTiming;
+  }
+}
+
+async function assertContentHashMatchesAsync(name: string, data: ArrayBuffer, expectedHashPrefix: string) {
+  const actualHashBuffer = await crypto.subtle.digest('SHA-256', data);
+  const actualHash = new Uint8Array(actualHashBuffer);
+  for (var byteIndex = 0; byteIndex*2 < expectedHashPrefix.length; byteIndex++) {
+    const expectedByte = parseInt(expectedHashPrefix.substr(byteIndex * 2, 2), 16);
+    const actualByte = actualHash[byteIndex];
+    if (actualByte !== expectedByte) {
+      const actualHashString = Array.from(actualHash).map(b => b.toString(16).padStart(2, '0')).join('');
+      throw new Error(`Resource hash mismatch for '${name}'. Expected prefix: '${expectedHashPrefix}'. Actual hash: '${actualHashString}'`);
```

Comment:
Should this produce a more meaningful error? It doesn't help a developer figure out what's incorrect if they see this or what it's side-effects were.

Reply:
If you have suggestions I’m interested, but if this happens I’m not sure we have any more insight than the site owner about why they served mismatching resources.

I guess we could say something like “try reloading the page in case a deployment was in progress”.

Reply:
> I guess we could say something like “try reloading the page in case a deployment was in progress”.

That sounds more useful. I compare the current message to the one [we produce on Blazor server](https://github.com/dotnet/aspnetcore/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+The+list+of+component+records+is+not+valid). I don't think anybody outside of the team really would understand what to do if they see this error in their console.

Reply:
OK, done